components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/jwt/JWTBearerGrantHandler.java

```diff
@@ -0,0 +1,653 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License
+ */
+
+package org.wso2.carbon.identity.oauth2.token.handlers.grant.jwt;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.crypto.RSASSAVerifier;
+import com.nimbusds.jwt.ReadOnlyJWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.bouncycastle.asn1.eac.ECDSAPublicKey;
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.common.model.FederatedAuthenticatorConfig;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants;
+import org.wso2.carbon.identity.application.common.util.IdentityApplicationManagementUtil;
+import org.wso2.carbon.identity.oauth.cache.JWTCache;
+import org.wso2.carbon.identity.oauth.cache.JWTCacheEntry;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  Class to handle JSON Web Token(JWT) grant type
+ */
+public class JWTBearerGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    private static Log log = LogFactory.getLog(JWTBearerGrantHandler.class);
+    private JWTCache jwtCache;
+
+    // JWT Profile for Oauth2 constants
+    private static final String OAUTH_JWT_BEARER_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+    private static final String OAUTH_JWT_BEARER_GRANT_ERROR = "invalid_grant";
+    private static final String OAUTH_JWT_BEARER_ISSUER = "issuer";
+    private static final String OAUTH_JWT_ASSERTION = "assertion";
+
+    // allowed number of minutes a token can be old for
+    private static final int REJECT_TOKEN_ISSUED_BEFORE = 30;
```

Comment:
Would be able to find a better suiting variable name. eg: validaity_period. Also better to have this configurable.